server/src/main/java/org/jboss/as/server/controller/git/ElytronClientSshdSessionFactory.java

```diff
@@ -0,0 +1,226 @@
+
+                   for (CredentialItem i : credentialItems) {
+                       if (i instanceof CredentialItem.Password) {
+                           CredentialCallback callback = new CredentialCallback(SSHCredential.class, null);
+                           Callback[] callbacks = {callback};
+                           CLIENT.getCallbackHandler(config).handle(callbacks);
+                           SSHCredential credential = null;
+                           if (callback != null && callback.getCredential() != null) credential = callback.getCredential().castAs(SSHCredential.class);
+                           char[] password = credential.getPassphrase().castAndApply(PasswordCredential.class, c -> c.getPassword().castAndApply(ClearPassword.class, ClearPassword::getPassword));
+                           ((CredentialItem.Password) i).setValue(password);
+                           continue;
+                       }
+                   }
+               } catch (IOException  | UnsupportedCallbackException ex) {
+                   throw new UnsupportedCredentialItem(urIish, ex.getMessage());
+               }
+
+               return true;
+           }
+       });
+    }
+
+    @Override
+    @NonNull
+    protected List<Path> getDefaultKnownHostsFiles(@NonNull File sshDir) {
+        return Arrays.asList(sshDir.toPath().resolve("known_hosts"), sshDir.toPath().resolve(this.knownHostsFile));
+    }
+
+    @Override
+    @NonNull
+    protected Iterable<KeyPair> getDefaultKeys(@NonNull File sshDir) {
+        if(keyPair == null) {
+            List<Path> defaultIdentities = this.getDefaultIdentities(sshDir);
+            return (Iterable) this.defaultKeys.computeIfAbsent(new Tuple(defaultIdentities.toArray(new Path[0])), (t) -> {
+                return new CachingKeyPairProvider(defaultIdentities, this.getKeyCache());
+            });
+        } else {
+            return (Iterable) Arrays.asList(keyPair);
+        }
+    }
+
+    //Copied from org.eclipse.jgit.transport.sshd.SshdSessionFactory
+    private static final class Tuple {
+        private Object[] objects;
+
+        public Tuple(Object[] objects) {
+            this.objects = objects;
+        }
+
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            } else if (obj != null && obj.getClass() == Tuple.class) {
+                Tuple other = (Tuple)obj;
+                return Arrays.equals(this.objects, other.objects);
+            } else {
+                return false;
+            }
+        }
+
+        public int hashCode() {
+            return Arrays.hashCode(this.objects);
+        }
+    }
+
+    private URI urIishToUri(URIish urIish) throws URISyntaxException {
+        String scheme = "ssh";
+        String user = urIish.getUser();
+        String host = urIish.getHost();
+        int port = urIish.getPort();
+        String path = null;
+        if (user.equals("git")) {
```

Comment:
Would be good to add some comments here to explain what's happening.

Reply:
Fixed, thanks!